### PR TITLE
packaging: Add IBM Z SE artifacts to main

### PR DESCRIPTION
**Pull request by BbolroC**
This is to add artifacts for IBM Z SE(TEE) to main. 

This PR **MUST** be merged after https://github.com/kata-containers/kata-containers/pull/6586.

Fixes: #6754

Signed-off-by: Hyounggyu Choi <Hyounggyu.Choi@ibm.com>

### CODE REVIEW

File: .github/workflows/build-kata-static-tarball-s390x.yaml [hunk 1]
```diff
@@ -20,6 +20,7 @@ jobs:
           - qemu
           - rootfs-image
           - rootfs-initrd
+          - se-image
```
Comment (wainersm) [inline .github/workflows/build-kata-static-tarball-s390x.yaml:23]: IIU the se-image is a repack of rootfs-initrd with some specific stuffs and generated as a ramdisk image. So could we call this new image as "rootfs-image-se" to follow the name pattern?

File: ci/lib.sh [hunk 2]
```diff
@@ -64,3 +64,86 @@ run_get_pr_changed_file_details()
...
+	chmod 600 "${parmfile}"
+
+	[ -n "${HKD_PATH:-}" ] || (echo >&2 "No host key document specified." && return 1)
+	cert_list=($(ls -1 $HKD_PATH))
+	declare hkd_options
+	eval "for cert in ${cert_list[*]}; do
+		hkd_options+=\"--host-key-document=\\\"\$HKD_PATH/\$cert\\\" \"
+	done"
+
+	command -v genprotimg > /dev/null 2>&1 || { apt update; apt install -y s390-tools; }
```
Comment (wainersm) [inline ci/lib.sh:126]: I suggest to not install stuffs in this function. IMO the calling script should ensure the required tools are installed in the system. Apart from that, the kata-deploy's Dockerfile for s390x already install the packages ([tools/packaging/kata-deploy/local-build/dockerbuild/Dockerfile](https://github.com/kata-containers/kata-containers/pull/6755/commits/c92c91130644e145f485d5273475e808039407a6#diff-08d854f59d53cefd732ccc081adc1b49e6279e0e9bb33218ac99bbd56589e922))

File: ci/lib.sh [hunk 3]
```diff
@@ -64,3 +64,86 @@ run_get_pr_changed_file_details()
...
+		"${extra_arguments}" \
+		"${hkd_options}" \
+		--output="${install_dest_dir}/kata-containers-secure.img" \
+		--image="${install_src_dir}/vmlinuz.container" \
+		--ramdisk="${install_src_dir}/kata-containers-initrd.img" \
+		--parmfile="${parmfile}" \
+		--no-verify # no verification for CI testing purposes
+
+	build_result=$?
+	rm -f "${parmfile}"
```
Comment (wainersm) [inline ci/lib.sh:143]: This line is never executed if the check on line 119 failed. I suggest handling the paramfile removal within a "trap EXIT" instructions. Something like: `trap "rm -f ${paramfile}" EXIT`

**Comment by wainersm** [comment]
hi @BbolroC , thanks for helping on the CCv0 -> main merge! 

I left a few comments. Overall it looks good but it would be better if you could break it into smaller commits. Could you do that?

File: src/runtime/config/configuration-qemu-se.toml.in [hunk 4]
```diff
@@ -0,0 +1,673 @@
+# Copyright (c) 2017-2019 Intel Corporation
+# Copyright (c) 2021 Adobe Inc.
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+# XXX: WARNING: this file is auto-generated.
+# XXX:
+# XXX: Source file: "@CONFIG_QEMU_IN@"
```
Comment (fitzthum) [inline src/runtime/config/configuration-qemu-se.toml.in:9]: Should this point to `CONFIG_QEMU_SE_IN`?

File: ci/lib.sh [hunk 5]
```diff
@@ -64,3 +64,86 @@ run_get_pr_changed_file_details()
...
+#
+# Parameters:
+#	$1	- kernel_parameters
+#	$2	- a source directory where kernel and initrd are located
+#	$3	- a destination directory where a SE image is built
+#
+# Return:
+# 	0 if the image is successfully built
+#	1 otherwise
+build_secure_image() {
```
[the remark this answers is not in the document]
Comment (fitzthum) [reply ci/lib.sh:99]: in addition to location, maybe would be clearer to call this `build_secure_execution_image`

File: .github/workflows/build-kata-static-tarball-s390x.yaml [hunk 6]
```diff
@@ -31,6 +32,14 @@ jobs:
         with:
           ref: ${{ github.event.pull_request.head.sha }}
           fetch-depth: 0 # This is needed in order to keep the commit ids history
+
+      - name: Place a host key document
+        run: |
+          mkdir -p "host-key-document"
+          cp "${CI_HKD_PATH}" "host-key-document"
+        env:
+          CI_HKD_PATH: ${{ secrets.CI_HKD_PATH }}
```
Comment (fitzthum) [inline .github/workflows/build-kata-static-tarball-s390x.yaml:41]: So the HKD is specific to the machine where the image will be run, correct? For the CI, we have a secret that stores an HKD that matches the runner where we test things. What about the image that is deployed via the operator? Can this tooling provide a valid image built ahead of time or does the client have to build their own?

File: ci/lib.sh [hunk 7]
```diff
@@ -64,3 +64,86 @@ run_get_pr_changed_file_details()
...
+Either kernel or initrd does not exist or is mistakenly named
+A file name for kernel must be vmlinuz.container (raw binary)
+A file name for initrd must be kata-containers-initrd.img
+EOF
+		return 1
+	fi
+
+	cmdline="${kernel_params} panic=1 scsi_mod.scan=none swiotlb=262144"
+	parmfile="$(mktemp --suffix=-cmdline)"
+	echo "${cmdline}" > "${parmfile}"
```
Comment (fitzthum) [inline ci/lib.sh:116]: What is `parmfile`? Should it be `paramfile`?

File: tools/packaging/guest-image/build_se_image.sh [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+	image_source_dir="${builddir}/secure-image"
+	mkdir -p "${image_source_dir}"
+	pushd "${tarball_dir}"
+	for tarball_id in kernel rootfs-initrd; do
+		tar xvf kata-static-${tarball_id}.tar.xz -C "${image_source_dir}"
+	done
+	popd
+
+	protimg_source_dir="${image_source_dir}${prefix}/share/kata-containers"
+	local kernel_params="agent.enable_signature_verification=false"
```
Comment (fitzthum) [inline tools/packaging/guest-image/build_se_image.sh:50]: I don't think this feature is in main yet.

**Review by fitzthum — COMMENTED** [review]
A few comments

File: .github/workflows/build-kata-static-tarball-s390x.yaml (hunk 1 above)
Comment (BbolroC) [reply .github/workflows/build-kata-static-tarball-s390x.yaml:23]: Unfortunately, it is a repackage of kernel + rootfs-initrd. So it is a little brave to call it `rootfs-`.

File: ci/lib.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (BbolroC) [reply ci/lib.sh:99]: yeah, I will move it to the location you are recommending. For the naming, a `secure image` is used in the [official manual](https://public.dhe.ibm.com/software/dw/linux390/docu/l151se00.pdf). But I would like to hear another opinion from @hbrueckner. Thanks.

File: ci/lib.sh (hunk 2 above)
Comment (BbolroC) [reply ci/lib.sh:126]: yes, that is done by me. I will simply place the tool checking here. Thanks.

File: ci/lib.sh (hunk 3 above)
Comment (BbolroC) [reply ci/lib.sh:143]: I will update the code accordingly. Thanks.

**Comment by BbolroC** [comment]
> hi @BbolroC , thanks for helping on the CCv0 -> main merge!
> 
> I left a few comments. Overall it looks good but it would be better if you could break it into smaller commits. Could you do that?

Yes, I will follow your suggestion. Thanks a lot!

File: src/runtime/config/configuration-qemu-se.toml.in (hunk 4 above)
Comment (BbolroC) [reply src/runtime/config/configuration-qemu-se.toml.in:9]: Correct! Thanks. :wink:

File: .github/workflows/build-kata-static-tarball-s390x.yaml (hunk 1 above)
Comment (wainersm) [reply .github/workflows/build-kata-static-tarball-s390x.yaml:23]: ah, I see... indeed calling it `rootfs` would be wrong. What about `boot-image-se`? According to the doc you pasted below and IIUC this is referred as "boot image". Sorry for the nit, it is that "image" can be anything :D

File: .github/workflows/build-kata-static-tarball-s390x.yaml (hunk 6 above)
Comment (BbolroC) [reply .github/workflows/build-kata-static-tarball-s390x.yaml:41]: > So the HKD is specific to the machine where the image will be run, correct? For the CI, we have a secret that stores an HKD that matches the runner where we test things.

yes, correct.

> What about the image that is deployed via the operator? Can this tooling provide a valid image built ahead of time or does the client have to build their own?

ATM, a secure image in the operator is built only for a CI purpose. If a user wants to build an image and deploy it via the operator, he/she should provide his/her authenticated/authorized image registry (not sure if it is configurable on `main`) and an env variable `CI_HKD_PATH` where his/her own HKD is placed, which is bound to a machine to be deployed. 

Yes, this PR takes only a CI case into account.

File: ci/lib.sh (hunk 7 above)
Comment (BbolroC) [reply ci/lib.sh:116]: It follows the term at https://www.ibm.com/docs/en/linux-on-systems?topic=commands-genprotimg. But, yes, it means `parameters_file`.

File: tools/packaging/guest-image/build_se_image.sh (hunk 8 above)
Comment (BbolroC) [reply tools/packaging/guest-image/build_se_image.sh:50]: Then, we could start with an empty string. Thanks for the info.

File: .github/workflows/build-kata-static-tarball-s390x.yaml (hunk 1 above)
Comment (BbolroC) [reply .github/workflows/build-kata-static-tarball-s390x.yaml:23]: It sounds much better. Wdyt, @hbrueckner ?

File: .github/workflows/build-kata-static-tarball-s390x.yaml (hunk 6 above)
Comment (fitzthum) [reply .github/workflows/build-kata-static-tarball-s390x.yaml:49]: Ok, makes sense. Do we have a guide for clients who are building their own images?

Comment (BbolroC) [reply .github/workflows/build-kata-static-tarball-s390x.yaml:49]: That is a plan discussed with Hendrik today :wink:

File: tools/packaging/kata-deploy/local-build/Makefile [hunk 9]
```diff
@@ -9,6 +9,24 @@ MK_DIR := $(dir $(MK_PATH))
 # Verbose build
 V := 1
 
+ARCH := $(shell uname -m)
```
Comment (fitzthum) [inline tools/packaging/kata-deploy/local-build/Makefile:12]: I think we're now building the artifacts for the CI in VMs, so I think this ARCH will always be x86_64. I guess this conditional won't hurt. All the SE components can be cross-compiled, right?

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 10]
```diff
@@ -98,6 +99,7 @@ options:
 	shim-v2
 	tdvf
 	virtiofsd
+	boot-image-se
```
Comment (fitzthum) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:102]: I think this is supposed to be alphabetized.

**Review by fitzthum — COMMENTED** [review]
A few more comments.

File: .github/workflows/build-kata-static-tarball-s390x.yaml [hunk 11]
```diff
@@ -20,6 +20,7 @@ jobs:
           - qemu
           - rootfs-image
           - rootfs-initrd
+          - boot-image-se
```
[the remark this answers is not in the document]
Comment (BbolroC) [reply .github/workflows/build-kata-static-tarball-s390x.yaml:23]: Oh, I didn't notice that. Thanks.

File: src/runtime/Makefile [hunk 12]
```diff
@@ -350,6 +363,8 @@ ifneq (,$(QEMUCMD))
     KERNELSNPNAME = $(call MAKE_KERNEL_SNP_NAME,$(KERNELSNPTYPE))
     KERNELSNPPATH = $(KERNELDIR)/$(KERNELSNPNAME)
 
+    KERNELSENAME = kata-containers-secure.img
```
[the remark this answers is not in the document]
Comment (BbolroC) [reply src/runtime/Makefile:366]: You are right. I will go with `kata-containers-se-img` then.

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries-in-docker.sh [hunk 13]
```diff
@@ -54,6 +54,8 @@ docker run \
 	-v "${kata_dir}:${kata_dir}" \
 	--env CI="${CI:-}" \
 	--env USER=${USER} \
+	--env AA_KBC="${AA_KBC:-}" \
```
[the remark this answers is not in the document]
Comment (BbolroC) [reply tools/packaging/kata-deploy/local-build/kata-deploy-binaries-in-docker.sh:57]: Thanks for the info.

File: src/runtime/config/configuration-qemu-se.toml.in [hunk 14]
```diff
@@ -0,0 +1,673 @@
...
+# Enable debug console.
+
+# If enabled, user can connect guest OS running inside hypervisor
+# through "kata-runtime exec <sandbox-id>" command
+
+#debug_console_enabled = true
+
+# Agent connection dialing timeout value in seconds
+# (default: 30)
+#dial_timeout = 30
```
[the remark this answers is not in the document]
Comment (BbolroC) [reply src/runtime/config/configuration-qemu-se.toml.in:520]: Ah, good to know. I will update the value. Thanks.

File: tools/packaging/kata-deploy/local-build/Makefile (hunk 9 above)
Comment (BbolroC) [reply tools/packaging/kata-deploy/local-build/Makefile:12]: Yes, it is in progress (https://github.com/kata-containers/kata-containers/pull/6586). I have to change the conditional.

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh (hunk 10 above)
Comment (BbolroC) [reply tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:102]: Yep, thanks.

**Comment by BbolroC** [comment]
This PR should be merged after https://github.com/kata-containers/kata-containers/pull/6586 due to the cross-compilation.

File: tools/packaging/kata-deploy/local-build/Makefile [hunk 15]
```diff
@@ -9,6 +9,26 @@ MK_DIR := $(dir $(MK_PATH))
 # Verbose build
 V := 1
 
+ifeq ($(CROSS_BUILD),false)
```
Comment (wainersm) [inline tools/packaging/kata-deploy/local-build/Makefile:12]: Hi @BbolroC ! The default behavior is the user not export `CROSS_BUILD=false` and this `ifeq` will fail, resulting on a wrong build for x86_64. You should set `CROSS_BUILD=false` if not exported.

Also I suggest to set `EXTRA_TARBALL` and `SERIAL_EXTRA_TARBALL` with the x86_64 targets by default. If `CROSS_BUILD=true` and `ARCH=s390x` then overwrite `EXTRA_TARBALL` and `SERIAL_EXTRA_TARBALL`.

**Comment by wainersm** [comment]
On my workstation (x86_64) I rebased this PR on top of https://github.com/kata-containers/kata-containers/pull/6586 (enable cross build for non-x86) then I ran the following command to build the SE boot image:

```
CROSS_BUILD=true TARGET_ARCH=s390x ARCH=s390x make boot-image-se-tarball
```

It failed to build the kernel (pre-req of the boot image):

```
<OUTPUT OMITTED>

#7 naming to quay.io/kata-containers/builders:kernel-7b62aa69333117c07a066af8dd95c772fbeba5e1-x86_64-s390x-cross-build 0.0s done
#7 DONE 2.1s
INFO: Config version: 106
INFO: Kernel version: 5.19.2
INFO: kernel path does not exist, will download kernel
INFO: Download kernel checksum file: sha256sums.asc from https://cdn.kernel.org/pub/linux/kernel/v5.x/sha256sums.asc
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  292k  100  292k    0     0   158k      0  0:00:01  0:00:01 --:--:--  158k
INFO: Download kernel version 5.19.2
INFO: Download kernel
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   162  100   162    0     0    115      0  0:00:01  0:00:01 --:--:--   115
100  125M  100  125M    0     0   252k      0  0:08:28  0:08:28 --:--:--  287k
linux-5.19.2.tar.xz: OK
INFO: Apply patches from /home/wmoschet/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/patches/5.19.x
INFO: Found 0 patches
INFO: Constructing config from fragments: /home/wmoschet/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/configs/fragments/s390/.config
INFO: Some CONFIG elements failed to make the final .config:
INFO: Value requested for CONFIG_HAVE_EBPF_JIT not in final .config
INFO: Generated config file can be found in /home/wmoschet/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/configs/fragments/s390/.config
ERROR: Failed to construct requested .config file
ERROR: failed to find default config 
make[1]: *** [tools/packaging/kata-deploy/local-build/Makefile:59: kernel-tarball-build] Error 1
make[1]: Leaving directory '/home/wmoschet/go/src/github.com/kata-containers/kata-containers'
make: *** [tools/packaging/kata-deploy/local-build/Makefile:83: kernel-tarball] Error 2
```

Have I done it wrong?

**Comment by BbolroC** [comment]
> Have I done it wrong?

It looks like a common issue for the latest kernel. `CONFIG_HAVE_EBPF_JIT` is defined at [tools/packaging/kernel/configs/fragments/common/network.conf](https://github.com/kata-containers/kata-containers/blob/main/tools/packaging/kernel/configs/fragments/common/network.conf#L67), but the option looks to get dropped at the kernel.

I got the kernel compiled successfully by putting `CONFIG_HAVE_EBPF_JIT` into [tools/packaging/kernel/configs/fragments/whitelist.conf](https://github.com/kata-containers/kata-containers/blob/main/tools/packaging/kernel/configs/fragments/whitelist.conf)

**Review by fitzthum — APPROVED** [review]
I have not tried building the components, but the changes look good to me.

**Comment by BbolroC** [comment]
Then, I will let you know a result the minute I finish builing the component on my test machine. Thanks @wainersm and @fitzthum !

**Comment by BbolroC** [comment]
With a patch in https://github.com/kata-containers/kata-containers/pull/6586#issuecomment-1563164993, I've managed to get `kernel-tarball` and `rootfs-initrd-tarball` built successfully on my x86_64 test machine. I will test this PR on top of the patched #6586 next week. Have a lovely weekend! :wink:

**Comment by BbolroC** [comment]
@wainersm @fitzthum : I've verified to successfully build a secure image with the following steps after rebasing this onto #6586:

```
$ # At the project root
$ mkdir host-key-document
$ # Place a host key document (a key for encrypting an image, e.g. cp your_host_key host-key-document/)
$ export HKD_PATH=host-key-document
$ sudo -E PATH=$PATH CROSS_BUILD=true TARGET_ARCH=s390x ARCH=s390x make boot-image-se-tarball
... log skipped ...
Build kata version 3.2.0-alpha3: boot-image-se                                                                                                                                                                     INFO: DESTDIR /home/ubuntu/go/src/github.com/kata-containers/kata-containers/tools/packaging/kata-deploy/local-build/build/boot-image-se/destdir
INFO: Create IBM SE image configured with AA_KBC=                          
INFO: Build IBM zSystems & LinuxONE SE image                           
/home/ubuntu/go/src/github.com/kata-containers/kata-containers/tools/packaging/kata-deploy/local-build/build /home/ubuntu/go/src/github.com/kata-containers/kata-containers/tools/packaging/kata-deploy/local-build
/build/boot-image-se/builddir                         
./                                                         
./opt/                                                                           
./opt/kata/                                                                      
./opt/kata/share/                                                                                      
./opt/kata/share/kata-containers/                                                          
./opt/kata/share/kata-containers/vmlinuz.container                                     
./opt/kata/share/kata-containers/vmlinux-5.19.2-107                                 
./opt/kata/share/kata-containers/config-5.19.2                                          
./opt/kata/share/kata-containers/vmlinux.container
./opt/kata/share/kata-containers/vmlinuz-5.19.2-107    
./                                   
./opt/                                                 
./opt/kata/                                    
./opt/kata/share/                                               
./opt/kata/share/kata-containers/                            
./opt/kata/share/kata-containers/kata-containers-initrd.img           
./opt/kata/share/kata-containers/kata-ubuntu-20.04.initrd             
/home/ubuntu/go/src/github.com/kata-containers/kata-containers/tools/packaging/kata-deploy/local-build/build/boot-image-se/builddir
WARNING: host-key document verification is disabled. Your workload is not secured.
./                                                                                         
./opt/                                                                                                                                                                                                             
./opt/kata/                                                                                                                                                                                                        
./opt/kata/share/                                            
./opt/kata/share/kata-containers/                                          
./opt/kata/share/kata-containers/kata-containers-se.img                
drwxr-xr-x root/root         0 2023-06-01 20:20 ./          
drwxr-xr-x root/root         0 2023-06-01 20:20 ./opt/
drwxr-xr-x root/root         0 2023-06-01 20:20 ./opt/kata/
drwxr-xr-x root/root         0 2023-06-01 20:20 ./opt/kata/share/                
drwxr-xr-x root/root         0 2023-06-01 20:20 ./opt/kata/share/kata-containers/
-rw-r--r-- root/root  36958208 2023-06-01 20:20 ./opt/kata/share/kata-containers/kata-containers-se.img
make[1]: Leaving directory '/home/ubuntu/go/src/github.com/kata-containers/kata-containers'
```

FYI: I've added another commit for the adjustment for the cross-compilation.
I will raise an issue for providing a document like [what SNP people did ](https://github.com/kata-containers/kata-containers/blob/main/docs/how-to/how-to-run-kata-containers-with-SNP-VMs.md) for SE. Thanks for your patience.

**Comment by BbolroC** [comment]
Since #6586 was merged, I've resolved the conflicts and confirmed that an artifact `kata-containers-se.img` is built on a `x86_64` machine. It would be grateful to get this PR reviewed by @wainersm and @fidencio . Thanks!

File: .github/workflows/build-kata-static-tarball-s390x.yaml (hunk 6 above)
Comment (BbolroC) [reply .github/workflows/build-kata-static-tarball-s390x.yaml]: FYI: https://github.com/kata-containers/kata-containers/pull/7146 will cover a client's needs.

**Review by jodh-intel — APPROVED** [review]
Thanks @BbolroC.

File: tools/packaging/kata-deploy/local-build/Makefile [hunk 16]
```diff
@@ -9,19 +9,16 @@ MK_DIR := $(dir $(MK_PATH))
...
-define BUILD
-	$(MK_DIR)/kata-deploy-binaries-in-docker.sh $(if $(V),,-s) --build=$1
-endef
+ifeq ($(CROSS_BUILD),)
+	CROSS_BUILD = false
+endif
 
-kata-tarball: | all-parallel merge-builds
+ifeq ($(CROSS_BUILD),false)
+	ARCH := $(shell uname -m)
```
Comment (wainersm) [inline tools/packaging/kata-deploy/local-build/Makefile:17]: @BbolroC suppose that I want to build the SE rootfs on my x86_64 laptop. I export `CROSS_BUILD=true` and then should I export `ARCH=s390x`?

I didn't see in details the upcoming commits, but, should I also export `TARGET_ARCH=s390x` too?

Comment (BbolroC) [reply tools/packaging/kata-deploy/local-build/Makefile:17]: The following three are required:

```
CROSS_BUILD=true TARGET_ARCH=s390x ARCH=s390x
```

This is something we agreed on at https://github.com/kata-containers/kata-containers/pull/6586

**Review by wainersm — APPROVED** [review]
hi @BbolroC ! Sorry for taking months to re-review it. I couldn't build the boot-image-se due the lack of host key document on my laptop, then I reviewed the code carefully and didn't find any problem. The shim-v2, on the other hand, I could build and I see the configuration-qemu-se.toml generated and it looks good.

**Pull request merged**